### PR TITLE
[Lock] Use platform to identify the PDO driver

### LINKED ISSUE #43048: PostgreSqlStore breaks Sentry usage due to hard-coded instanceof checks.
PostgreSqlStore of any version up to 6.0 has a hard-coded switch in it's code [here](https://github.com/symfony/lock/blob/5.3/Store/PostgreSqlStore.php#L260):
```switch (true) {
                case $driver instanceof \Doctrine\DBAL\Driver\PDOPgSql\Driver:
                case $driver instanceof \Doctrine\DBAL\Driver\PDO\PgSQL\Driver:
                    break;
                default:
                    throw new InvalidArgumentException(sprintf('The adapter "%s" does not support the "%s" driver.', __CLASS__, \get_class($driver)));
            }
```

So you just get an exception "The adapter "Symfony\Component\Lock\Store\PostgreSqlStore" does not support the "Sentry\SentryBundle\Tracing\Doctrine\DBAL\TracingDriver" driver." when using Sentry with PostgreSqlStore.

There is a similar issue with Symfony cache [here](https://github.com/symfony/symfony/pull/42011). Filed Sentry issue here also: https://github.com/getsentry/sentry-symfony/issues/558

### PR DESCRIPTION
**Pull request by Jean85**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #43048
| License       | MIT

This should fix the issue in the same way we did for #42011. I'm not sure if I should add/change any test...

### CODE REVIEW

File: src/Symfony/Component/Lock/Store/PdoStore.php [hunk 1]
```diff
@@ -332,34 +332,17 @@ private function getDriver(): string
...
-                case $driver instanceof \Doctrine\DBAL\Driver\SQLSrv\Driver:
-                case $driver instanceof \Doctrine\DBAL\Driver\PDOSqlsrv\Driver:
-                case $driver instanceof \Doctrine\DBAL\Driver\PDO\SQLSrv\Driver:
-                    $this->driver = 'sqlsrv';
-                    break;
-                default:
-                    $this->driver = \get_class($driver);
-                    break;
+            if ($driver instanceof \Doctrine\DBAL\Driver\Mysqli\Driver) {
+                throw new \LogicException(sprintf('The adapter "%s" does not support the mysqli driver, use pdo_mysql instead.', static::class));
```
Comment (stof) [inline src/Symfony/Component/Lock/Store/PdoStore.php:337]: why is it unsupported ? That looks suspicious to me. We already have the logic in place to use SQL compatible with MySQL, and I don't think we go low-level enough in our usage of DBAL APIs to be concerned about other kind of differences between drivers than the platform (which requires escaping the abstraction layer)

Comment (jderusse) [reply src/Symfony/Component/Lock/Store/PdoStore.php:337]: for reference https://github.com/symfony/symfony/pull/35375

**Review by jderusse — COMMENTED** [review]
This PR revert what have been done in https://github.com/symfony/symfony/commit/7eca3a59700b3925acd03d05858beb61464c4cc3#diff-98857ac5bb047b333495db5092ea4b875924022056c5b5522d7ecb224fac5f6f

@nicolas-grekas  do you remember why you changed this in a merge commit?

File: src/Symfony/Component/Lock/Store/PdoStore.php [hunk 2]
```diff
@@ -332,34 +332,17 @@ private function getDriver(): string
...
             }
+
+            $this->driver = [
+                'mssql' => 'sqlsrv',
+                'oracle' => 'oci',
+                'postgresql' => 'pgsql',
+                'sqlite' => 'sqlite',
+                'mysql' => 'mysql',
+            ][$driver->getDatabasePlatform()->getName()] ?? \get_class($driver);
```
Comment (derrabus) [inline src/Symfony/Component/Lock/Store/PdoStore.php]: The `getName()` method has been deprecated recently. We should work with `instanceof` instead.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Lock/Store/PdoStore.php (hunk 2 above)
Comment (Jean85) [reply src/Symfony/Component/Lock/Store/PdoStore.php]: It has been deprecated in 3.2 (which is unreleased IIRC); we support 2.x and 3.x, are we sure that we can rely on that extensively?

Comment (stof) [reply src/Symfony/Component/Lock/Store/PdoStore.php]: Doing `$driver->getDatabasePlatform() instanceof MysqlPlatform` (or similar checks) works in both 2.x and 3.x

Comment (derrabus) [reply src/Symfony/Component/Lock/Store/PdoStore.php]: We can leave it as is for now, if you prefer. It's just that we will need to change this piece of code again sooner or later.

**Review by derrabus — APPROVED** [review]

**Review by jderusse — COMMENTED** [review]
Regarding the @derrabus  comment, this PR ~brings nothing but~ introduce a future deprecation for upcoming Doctrine 3.2

I'm 👎 for this change

edit: this PR fixes the issue referenced in the description

**Comment by stof** [comment]
@jderusse this PR fixes compatibility with driver decorators (inspecting the driver class to find the target platform is a mistake). See the ticket referenced in the issue description, which describes precisely such a bug.

**Comment by jderusse** [comment]
> @jderusse this PR fixes compatibility with driver decorators (inspecting the driver class to find the target platform is a mistake). See the ticket referenced in the issue description, which describes precisely such a bug.

Still.. `getName` will be removed in the next major version (https://github.com/doctrine/dbal/pull/4763) fixing the issue by introducing a dependency with a (soon) deprecated method is not the right way to fix the issue IMHO.

**Comment by chalasr** [comment]
I agree, seems worth trying to fix this using `instanceof` checks on the platform instance instead so we don't have to revisit this too soon.

**Comment by Jean85** [comment]
I've changed the approach to use the `instanceof` on the platforms.

File: src/Symfony/Component/Lock/Store/PdoStore.php [hunk 3]
```diff
@@ -330,19 +330,32 @@ private function getDriver(): string
         if ($con instanceof \PDO) {
             $this->driver = $con->getAttribute(\PDO::ATTR_DRIVER_NAME);
         } else {
-            $driver = $con->getDriver();
```
Comment (fabpot) [inline src/Symfony/Component/Lock/Store/PdoStore.php:333]: Keeping this temp variable would make the diff a bit simpler and might ease merging from older branches.

File: src/Symfony/Component/Lock/Store/PdoStore.php [hunk 4]
```diff
@@ -157,7 +157,7 @@ public function waitAndSave(Key $key)
     public function putOffExpiration(Key $key, $ttl)
     {
         if ($ttl < 1) {
-            throw new InvalidTtlException(sprintf('"%s()" expects a TTL greater or equals to 1 second. Got %s.', __METHOD__, $ttl));
+            throw new InvalidTtlException(sprintf('"%s()" expects a TTL greater or equals to 1 second. Got "%s".', __METHOD__, $ttl));
```
Comment (fabpot) [inline src/Symfony/Component/Lock/Store/PdoStore.php]: Should be reverted (fabbot false positive)

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Component/Lock/Store/PdoStore.php [hunk 5]
```diff
@@ -331,33 +331,30 @@ private function getDriver(): string
...
                     $this->driver = 'mysql';
                     break;
-                case $driver instanceof \Doctrine\DBAL\Driver\PDOSqlite\Driver:
-                case $driver instanceof \Doctrine\DBAL\Driver\PDO\SQLite\Driver:
+                case $platform instanceof \Doctrine\DBAL\Platforms\SqlitePlatform:
                     $this->driver = 'sqlite';
                     break;
-                case $driver instanceof \Doctrine\DBAL\Driver\PDOPgSql\Driver:
-                case $driver instanceof \Doctrine\DBAL\Driver\PDO\PgSQL\Driver:
+                case $platform instanceof \Doctrine\DBAL\Platforms\PostgreSQLPlatform:
```
Comment (derrabus) [inline src/Symfony/Component/Lock/Store/PdoStore.php:348]:
```suggestion
                case $platform instanceof \Doctrine\DBAL\Platforms\PostgreSQLPlatform:
                case $platform instanceof \Doctrine\DBAL\Platforms\PostgreSQL94Platform:
```

For DBAL 3.1 😓

File: src/Symfony/Component/Lock/Store/PdoStore.php [hunk 6]
```diff
@@ -331,33 +331,30 @@ private function getDriver(): string
...
-                case $driver instanceof \Doctrine\DBAL\Driver\OCI8\Driver:
-                case $driver instanceof \Doctrine\DBAL\Driver\PDOOracle\Driver:
-                case $driver instanceof \Doctrine\DBAL\Driver\PDO\OCI\Driver:
+                case $platform instanceof \Doctrine\DBAL\Platforms\OraclePlatform:
                     $this->driver = 'oci';
                     break;
-                case $driver instanceof \Doctrine\DBAL\Driver\SQLSrv\Driver:
-                case $driver instanceof \Doctrine\DBAL\Driver\PDOSqlsrv\Driver:
-                case $driver instanceof \Doctrine\DBAL\Driver\PDO\SQLSrv\Driver:
+                case $platform instanceof \Doctrine\DBAL\Platforms\SQLServerPlatform:
```
Comment (derrabus) [inline src/Symfony/Component/Lock/Store/PdoStore.php:355]:
```suggestion
                case $platform instanceof \Doctrine\DBAL\Platforms\SQLServerPlatform:
                case $platform instanceof \Doctrine\DBAL\Platforms\SQLServer2012Platform:
```

**Review by derrabus — APPROVED** [review]

**Comment by derrabus** [comment]
Thank you @Jean85.

File: src/Symfony/Component/Lock/Store/PdoStore.php (hunk 5 above)
Comment (stof) [reply src/Symfony/Component/Lock/Store/PdoStore.php:348]: what ? They removed the base unversioned class ? that's quite bad newws

Comment (Jean85) [reply src/Symfony/Component/Lock/Store/PdoStore.php:348]: Reverted back in 3.2, and deprecated the whole "platform classes for specific versions": https://github.com/doctrine/dbal/commit/666c0656c03587802b1524a9b54e8a104a6f03b1

**Pull request merged**